src/Common/src/Microsoft/Win32/SafeHandles/SafeHandleCache.cs

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Threading;
+
+namespace Microsoft.Win32.SafeHandles
+{
+    /// <summary>Provides a cache for special instances of SafeHandles.</summary>
+    /// <typeparam name="T">Specifies the type of SafeHandle.  The type must have a parameterless constructor.</typeparam>
+    internal static class SafeHandleCache<T> where T : SafeHandle
+    {
+        private static T s_invalidHandle;
+
+        /// <summary>
+        /// Gets a cached, invalid handle.  As the instance is cached, it should either never be Disposed
+        /// or it should override <see cref="SafeHandle.Dispose(bool)"/> to prevent disposal when the
+        /// instance is the <see cref="InvalidHandle"/>.
+        /// </summary>
+        internal static T InvalidHandle
+        {
+            get
+            {
+                T currentHandle = Volatile.Read(ref s_invalidHandle);
```

Comment:
> provides no benefit

I think there is a benefit: t highlights the intent, and it prevents the unnecessary instantiation in the case you mention.  Now, those benefits may be very small, but so too is any performance impact.  Across a billion invocations, run repeatedly, I couldn't find any measurably difference between the two.  Are you seeing otherwise?